docs(adr): ADR-004: Expiration of application links (#627)
desmos-labs/desmos: docs/architecture/adr-004-application-links-expiration.md

## Context

Desmos `profiles` module give the possibility to link the desmos profile with both centralized application and 
other blockchains addresses. By doing this, a user can be verified as the owner of those accounts and prove to the system
that he's not impersonating anyone else. This verification however remains valid only if the user
never trade/sell his centralized-app username to someone else. If it does, and he already linked his Desmos profile
with some application, the link should be invalidated. Unfortunately for us, it's not possible to understand
when this happens since it's off-chain. To prevent this situation, an "expiration time" SHOULD be added 
to the centralized `app_links` object.

## Decision

First we expand the `Application link` structure by including a `expiration_height` field:  
```protobuf
message ApplicationLink {
  [...]
  // ExpirationBlockHeight represents the block height in which the link will be expired
  uint64 expiration_block_height = 7 [ (gogoproto.moretags) = "yaml:\"expiration_block_height\"" ];
}
```

> Then we add a new `ApplicationLinkState` `EXPIRED` necessary to address the no-longer valid links.

Author: Yes, your right

# ADR 004: Expiration of application links

## Changelog

- September 20th, 2021: Initial draft;
- September 21th, 2021: Moved from DRAFT to PROPOSED;
- December  22th, 2021: First review;
- January   04th, 2022: Second review.

## Status

PROPOSED

## Abstract

> Currently when a user links their centralized application with the Desmos profile, the created link contains a timestamp of when it has been created.   
> Since centralized applications' username can be switched and sold, we SHOULD implement an "expiration date" system on links. 
> This means that after a certain amount of time passes, the link will be automatically marked as invalid, and the user has to connect it again in order to keep it valid.

Reviewer: ```suggestion
Currently when a user links a centralized application with their Desmos profile, the created link contains a timestamp of when it has been created.   
Since centralized applications' username can be switched and sold, we SHOULD implement an "expiration date" system on links. This means that after a certain amount of time passes, the link will be automatically deleted and the user has to connect it again in order to keep it valid.
```

## Context

> Desmos `profiles` module give the possibility to link the desmos profile with both centralized application and 
> other blockchains accounts. By doing this, a user can be verified as the owner of those accounts and prove to the system
> that they're not impersonating anyone else. This verification however remains valid only if the user
> never trade or sell his centralized-app username to someone else. If they do the link to such username MUST be invalidated. 
> Unfortunately for us, it's not possible to understand when this happens since it's off-chain. 
> To prevent this situation, an "expiration time" SHOULD be added to the centralized `ApplicationLink` object.

Reviewer: ```suggestion
Desmos `profiles` module gives users the possibility to link their Desmos profile with both centralized applications and 
other blockchains accounts. By doing this, a user can be verified as the owner of those accounts and prove to the system that they're not impersonating anyone else. This verification however remains valid only if the user
never trades or sells their centralized app username to someone else. If they do, the link to such username MUST be invalidated. 
Unfortunately for us, it's not possible to understand when this happens since it's off-chain. 
To prevent this situation, an "expiration time" SHOULD be added to the `ApplicationLink` object.
```

> ## Decision

Reviewer: I'm honestly not a big fan of all this code inside the `Decisions` section. The more code we include, the more probabilities there are that we later need to change this ADR if we find better alternatives during its implementation. 

For this reason, I think we should instead describe what we should do on a high level. Something like this: 

```
In order to implement application link expiration, we will act as follow: 
1. Extend the `ApplicationLink` object by adding an expiration time field, which identifies the exact time on which the link will expire and thus SHOULD be deleted from the store 
2. Create a new `Keeper` method that is going to be called at the beginning of each block and delete all the expired application links. 

Also, in order to provide an easy way for the community to choose how long application links should be considered valid, we will also add the new params structure: 

...
```

    // Calculate the expiration time for the link
    expirationTimeParam := k.GetParams(ctx).AppLinks.ExpirationTime
    creationTime := ctx.BlockTime()
    expirationTime := creationTime.Add(time.Duration(expirationTimeParam.UnixNano()))

    // Store the connection
    err = k.SaveApplicationLink(ctx, types.NewApplicationLink(
            sender.String(),
            applicationData,
            types.ApplicationLinkStateInitialized,
            types.NewOracleRequest(
                0,
                oraclePrams.ScriptID,
                types.NewOracleRequestCallData(applicationData.Application, dataSourceCallData),
                clientID,
            ),
            nil,
            creationTime,
            expirationTime,
        ))
	[...]
}
```

## Consequences

### Backwards Compatibility

This update will affect the `ApplicationLink` object by adding the new `ExpirationBlockHeight` 
field on it breaking the compatibility with the previous versions of the software. To allow
a smooth update and overcome these compatibility issues, we need to set up a proper migration
from the previous versions to the one that will include the additions contained in this ADR.

```go

> func (m Migrator) Migrate(ctx sdk.Context) error {

Reviewer: We can delete this code as well

    params := m.keeper.GetParams(ctx)
    blockTime := ctx.BlockTime()
    expirationTimeParam := params.AppLinks.ExpirationTime

    m.keeper.IterateApplicationLinks(ctx, func(index int64, link types.ApplicationLink) (stop bool) {
        expirationTime := link.CreationTime.Add(time.Duration(expirationTimeParam.UnixNano()))
        // if the existent app link is expired, delete it
        if expirationTime.Before(blockTime) {
            m.keeper.deleteApplicationLinkStoreKeys(ctx, link)
            return false
        }

        link.ExpirationTime = expirationTime
        // can this error be unchecked? it checks if the link is associated to a profile
        _ = m.keeper.SaveApplicationLink(ctx, link)
        return false
    })

    return nil
}
```

### Positive

* Considerably reduce the possibility of impersonation of entities and users of centralized apps;

### Negative

> * Some more operations to perform in the overall handling of the app links

Reviewer: I think this negative point should be better explained. By adding the above described checks we are adding a lot of complexity to the handling of each block: we need to iterate over all the links and find all those expired to delete them. This is not an easy task and can take a consistent amount of time if the links present are a lot (indeed, we should add some benchmark test)

Author: We're going to mitigate this complexity by iterating only over the expired links as showed in the ADR code, 
by saving a reference of all the expirations times of the links.

Currently when a user links a centralized application with their Desmos profile, the created link contains a timestamp of when it has been created.   
Since centralized applications' username can be switched and sold, we SHOULD implement an "expiration date" system on links. 
This means that after a certain amount of time passes, the link will be automatically marked deleted and the user has to connect it again in order to keep it valid.

## Context

Desmos `x/profiles` module gives users the possibility to link their Desmos profile with both centralized application and 
other blockchains accounts. By doing this, a user can be verified as the owner of those accounts and prove to the system
that they're not impersonating anyone else. This verification however remains valid only if the user
never trades or sells their centralized-app username to someone else. If they do, the link to such username MUST be invalidated. 
Unfortunately for us, it's not possible to understand when this happens since it's off-chain. 
To prevent this situation, an "expiration time" SHOULD be added to the `ApplicationLink` object.

## Decision

To implement the link expiration we will act as follow:

> 1. Extend the `ApplicationLink` structure by adding an `ExpiratonTime` field that represents the time when the link will expire and will be deleted from the store.
> 2. Save a reference of the expiring link inside the store using a prefix and a `time.Time` value which will make it easy to iterate over the expired links
> 3. Add a new keeper method that will allow to iterate over these expired links
> 4. Use the created iterator to implement another keeper method `DeleteExpiredApplicationLinks` that will take care of deleting all the expired
>    links
> 5. Add this method to the `BeginBlock` function in order to trigger the deletion of expired link at each block

Reviewer: ```suggestion
1. extend the `ApplicationLink` structure by adding an `ExpiratonTime` field that represents the time when the link will expire and will be deleted from the store;
2. save a reference of the expiring link inside the store using a prefix and a `time.Time` value which will make it easy to iterate over the expired links;
3. add a new keeper method that allows to iterate over the expired links;
4. inside the `BeginBlock` method, iterate over all the expired links and delete them from the store.
```

We will also need to introduce a new `x/profiles` on chain parameter named `AppLinkParams` which contains 
the default validity duration of all the app links. The parameter will be later used inside the `StartProfileConnection` 

> to calculate the estimated expire time of each `AppLink`.

Reviewer: ```suggestion
to calculate the estimated expiration time of each `AppLink`.
```

## Consequences

### Backwards Compatibility

This update will affect the `ApplicationLink` object by adding the new `ExpirationTime` 
field and breaking the compatibility with the previous versions of the software. To allow
a smooth update and overcome these compatibility issues, we need to set up a proper migration
from the previous versions to the one that will include the additions contained in this ADR.

### Positive

- Considerably reduce the possibility of impersonation of entities and users of centralized apps;

### Negative

> By adding the extra `ExpirationTime` field we are going to raise the overall `AppLinks` handling complexity.
> Since we're performing an iteration over all the expired link references inside 
> at the start of each block this can require an amount of time that 
> SHOULD be studied with benchmark tests during the implementation.

Reviewer: ```suggestion
- By adding the extra `ExpirationTime` field we are going to raise the overall `AppLinks` handling complexity. Since we're performing an iteration over all the expired link references inside at the start of each block this can require an amount of time that SHOULD be studied with benchmark tests during the implementation.
```

# ADR 004: Expiration of application links

## Changelog

- September 20th, 2021: Initial draft;
- September 21th, 2021: Moved from DRAFT to PROPOSED;
- December  22th, 2021: First review;
- January   04th, 2022: Second review;
- January   10th, 2022: Third review;
- January   12th, 2022: Fourth review.

## Status

PROPOSED

## Abstract

Currently when a user links a centralized application with their Desmos profile, the created link contains a timestamp of when it has been created.   
Since centralized applications' username can be switched and sold, we SHOULD implement an "expiration date" system on links. 
This means that after a certain amount of time passes, the link will be automatically marked deleted and the user has to connect it again in order to keep it valid.

## Context

> Desmos `profiles` module gives users possibility to link their desmos profile with both centralized application and 

Reviewer: ```suggestion
Desmos `x/profiles` module gives users the possibility to link their Desmos profile with both centralized application and 
```

other blockchains accounts. By doing this, a user can be verified as the owner of those accounts and prove to the system
that they're not impersonating anyone else. This verification however remains valid only if the user
never trades or sells their centralized-app username to someone else. If they do, the link to such username MUST be invalidated. 
Unfortunately for us, it's not possible to understand when this happens since it's off-chain. 
To prevent this situation, an "expiration time" SHOULD be added to the `ApplicationLink` object.

## Decision

To implement the link expiration we will act as follow:

> 1. Extend the `ApplicationLink` structure by adding an `expiration_time` field that represent the time when the link will expire and will be

Reviewer: ```suggestion
1. Extend the `ApplicationLink` structure by adding an `ExpirationTime` field that represent the time when the link will expire and will be
```

> 1. Extend the `ApplicationLink` structure by adding an `expiration_time` field that represent the time when the link will expire and will be
>    deleted from the store.

Reviewer: ```suggestion
1. Extend the `ApplicationLink` structure by adding an `ExpiratonTime` field that represents the time when the link will expire and will be deleted from the store.
```

> 2. Save a reference of the expiring link inside the store under the prefix `ExpiringApplicationLinkPrefix`. It will be formed by a prefix and a 
>    `timestamp` value which will make it easy to iterate over the expired links

Reviewer: ```suggestion
2. Save a reference of the expiring link inside the store using a prefix and a `time.Time` value which will make it easy to iterate over the expired links
```

We also need to add a new `AppLinksParams` parameter to the `x/profiles` params, which will contain
an `expiration_time` timestamp that will be added to all the `AppLinks` before they're stored in the chain. 
The implementation for the new parameter is the following:

## Consequences

### Backwards Compatibility

This update will affect the `ApplicationLink` object by adding the new `ExpirationBlockHeight` 
field on it breaking the compatibility with the previous versions of the software. To allow
a smooth update and overcome these compatibility issues, we need to set up a proper migration
from the previous versions to the one that will include the additions contained in this ADR.

### Positive

* Considerably reduce the possibility of impersonation of entities and users of centralized apps;

### Negative

By adding the extra `expiration_time` field we are going to raise the overall `AppLinks` handling complexity.
Since we're performing an iteration over all the expired link references inside at the start of each block this can require
an amount of time that SHOULD be studied with benchmark tests during the implementation.

### Neutral

(none known)

## Further Discussions

> ## Test Cases [optional]

Reviewer: We should add a test here to make sure that expired application links are deleted when necessary
